drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+            .map_err(|_| SpCtrlError::Fault)?;
+        self.write_single_target_addr(DCRSR, r as u32 | (1u32 << 16))
+            .map_err(|_| SpCtrlError::Fault)?;
+
+        const RETRY_LIMIT: u32 = 10;
+        let mut limit = RETRY_LIMIT;
+        loop {
+            match self.dp_read_bitflags::<Dhcsr>() {
+                Ok(dhcsr) => {
+                    if dhcsr.is_regrdy() {
+                        // Trace retries used
+                        if limit != RETRY_LIMIT {
+                            ringbuf_entry!(Trace::LimitRemaining(limit));
+                        }
+                        return Ok(());
+                    }
+                    if limit == 0 {
+                        ringbuf_entry!(Trace::LimitRemaining(limit));
+                        return Err(SpCtrlError::Fault);
+                    }
+                    limit -= 1;
+                    hl::sleep_for(1);
+                }
+                Err(_) => return Err(SpCtrlError::Fault),
+            }
+        }
+    }
+
+    /// Measure the current SP Hubris Image.
+    /// The SP reset vector has been trapped and
+    /// the SP is halted.
+    fn do_measure_sp(&mut self) -> Result<[u8; 256 / 8], ()> {
+        // For Hubris on the STM32H7, The FWID includes 0xff padding to
+        // the end of the flash bank.
+
+        let mut magic = 0;
+        let mut total_image_size = 0;
+
+        // Sanity check the SP Hubris image
+        if self
+            .read_buf_from_addr(
+                endoscope::FLASH_BASE + sp::IMAGE_HEADER_MAGIC_OFFSET,
+                magic.as_bytes_mut(),
+            )
+            .is_err()
+            || magic != userlib::HEADER_MAGIC
+            || self
+                .read_buf_from_addr(
+                    endoscope::FLASH_BASE + sp::IMAGE_HEADER_LENGTH_OFFSET,
+                    total_image_size.as_bytes_mut(),
+                )
+                .is_err()
+            || !(sp::MIN_IMAGE_SIZE..(endoscope::FLASH_SIZE as usize))
+                .contains(&(total_image_size as usize))
+        {
+            ringbuf_entry!(Trace::SpHubrisFailsSanityCheck);
+            return Err(());
+        }
+
+        ringbuf_entry!(Trace::PlausibleSpHubris { total_image_size });
+
+        // Time the code injection injection, calculation,
+        // and readout of the FWID.
+        let start = sys_get_timer().now;
+        let r = self.sp_measure_fast();
+        let now = sys_get_timer().now;
+
+        ringbuf_entry!(Trace::MeasuredSp {
+            success: r.is_ok(),
+            delta_t: (now - start) as u32
+        });
+
+        r
+    }
+
+    fn read_buf_from_addr(
+        &mut self,
+        addr: u32,
+        buf: &mut [u8],
+    ) -> Result<(), SpCtrlError> {
+        let start = addr;
+        let end = addr + buf.len() as u32;
+        self.start_read_transaction(start, ((end - start) as usize) / 4)
+            .map_err(|_| SpCtrlError::Fault)?;
+
+        let cnt = buf.len();
+        if cnt % 4 != 0 {
```

Comment:
Nit: should we do this check _before_ starting a read transaction, to avoid leaving the SP in a weird state?

Reply:
This check is on the leased buffer. The calling task can mess this up and this is the place to deal with it. This PR doesn't try to change these APIs.

Reply:
I agree that the calling task can mess this up.  I think we should check for that condition before calling `start_read_transaction`.  Right now, if the calling task messes it up, then we start the read transaction _before_ returning an error, which leaves the SP in a weird state.